### PR TITLE
Date picker i18n

### LINKED ISSUE #12185: Localize showDatePicker
This issue is split from https://github.com/flutter/flutter/issues/10616 to avoid too much scope for one issue.

## Proposed scope

As part of this issue we will localize Gregorian calendar only. This will include:

 - [x] translations of numerals, week days, month names and buttons
 - [x] first day of week localization
 - [x] RTL
 - [x] move the `package:intl` dependency into `MaterialLocalizations`
 - [ ] tests

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes #12185. This version pull in all localizations provided by `package:intl`. I'm working on a separate PR that reduces `package:intl` to only the list of locales supported by Flutter.

### CODE REVIEW

File: packages/flutter/lib/src/material/date_picker.dart [hunk 1]
```diff
@@ -301,18 +321,64 @@ class DayPicker extends StatelessWidget {
...
+  /// - [MaterialLocalizations.firstDayOfWeekIndex] provides a 0-based index
+  ///   into the [MaterialLocalizations.narrowWeekDays] list.
+  /// - [MaterialLocalizations.narrowWeekDays] list provides localized names of
+  ///   days of week, always starting with Sunday and ending with Saturday.
+  int _computeFirstDayOffset(int year, int month, MaterialLocalizations localizations) {
+    // 0-based day of week, with 0 representing Monday.
+    final int weekDayFromMonday = new DateTime(year, month).weekday - 1;
+    // 0-based day of week, with 0 representing Sunday.
+    final int firstDayOfWeekFromSunday = localizations.firstDayOfWeekIndex;
+    // firstDayOfWeekFromSunday recomputed to be Sunday-based
```
Comment (Hixie) [inline packages/flutter/lib/src/material/date_picker.dart]: Monday-based?

File: packages/flutter/lib/src/material/date_picker.dart [hunk 2]
```diff
@@ -902,7 +978,31 @@ Future<DateTime> showDatePicker({
...
+      child: child,
+    );
+  }
+
+  if (textDirection != null) {
+    child = new Directionality(
+      textDirection: textDirection,
+      child: child,
+    );
+  }
```
Comment (Hixie) [inline packages/flutter/lib/src/material/date_picker.dart]: if you provide both, the way you have this built the locale will override the directionality. Presumably if both are given, the directionality should be on the inside.

**Comment by Hixie** [comment]
Can we fix https://github.com/flutter/flutter/issues/12050 first? Or at least next? I would rather we didn't keep adding dependencies on `intl`.

File: packages/flutter/test/material/date_picker_test.dart [hunk 3]
```diff
@@ -4,7 +4,9 @@
 
 import 'package:flutter/material.dart';
 import 'package:flutter_test/flutter_test.dart';
-import 'package:intl/intl.dart';
+import 'package:intl/intl.dart' hide TextDirection;
```
Comment (Hixie) [inline packages/flutter/test/material/date_picker_test.dart]: as intl

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/material/date_picker.dart (hunk 1 above)
Comment (yjbanov) [reply packages/flutter/lib/src/material/date_picker.dart]: LOL. See how confusing this is? 😄  Done. Good catch!

File: packages/flutter/lib/src/material/date_picker.dart (hunk 2 above)
Comment (yjbanov) [reply packages/flutter/lib/src/material/date_picker.dart]: Done, also actually used the `child` in the dialog! Also added tests.

File: packages/flutter/test/material/date_picker_test.dart (hunk 3 above)
Comment (yjbanov) [reply packages/flutter/test/material/date_picker_test.dart]: Actually, removed this import altogether. It was used only in one place, which can use `MaterialLocalizations` for this purpose.

**Comment by Hixie** [comment]
This had a bigger-than-usual negative impact on the aot_snapshot_size benchmarks.

**Comment by yjbanov** [comment]
> This had a bigger-than-usual negative impact on the aot_snapshot_size benchmarks.

This is partly because we added localized dates, and partly because we added localized dates for _all locales in the world_. https://github.com/flutter/flutter/pull/12466 partially addresses the issue.

Here are the numbers for building `complex_layout`:

`aot_snapshot_size_isolate`:

Before this PR: 1832972 bytes
After this PR: 1941640 bytes
After #12466: 1867322 bytes

`aot_snapshot_size_rodata`:

Before this PR: 1758568 bytes
After this PR: 1892776 bytes
After #12466: 1794616

**Pull request merged**